include/pybind11/pybind11.h

```diff
@@ -1064,24 +1064,68 @@ template <typename Type, typename... Extra> iterator make_iterator(Type &value,
     return make_iterator(std::begin(value), std::end(value), extra...);
 }
 
+NAMESPACE_BEGIN(detail)
+// These have to be separated like this because the lambda in the proper version won't compile if
+// the C++ types aren't implicitly convertible.
+template <typename InputType, typename OutputType>
+typename std::enable_if<std::is_convertible<InputType, OutputType>::value>::type
+implicitly_cpp_convertible(detail::type_info &input_type) {
+    // Some validity checks on output type:
+    if (!std::is_destructible<OutputType>::value)
+        pybind11_fail("implicitly_convertible: " + type_id<OutputType>() + " is not destructible");
+    if (!std::is_move_constructible<OutputType>::value && !std::is_copy_constructible<OutputType>::value)
```

Comment:
It's possible for OutputType to be movable but not copyable, but still support implicit conversion (see example below).  We prefer a move constructor, but if that isn't available, fallback to a copy constructor--so this is just checking that at least one of those is available.  This if statements (along with the one before it) should also be compiled away entirely by any decent compiler, since the ::value's are constexpr bools.



Example: implicit conversion of a movable, non-copyable type:
```
class In {};
class Out {
public:
    Out() = default;
    Out(const Out &) = delete;
    Out(Out &&) = default;
    Out(const In&) {}
};

void out(const Out&) { std::cout << "out called\n"; }

int main() {
    In in;
    out(in);
}
```